### DISCUSSION TITLE
Fields and wide menu #1

### DISCUSSION THREAD

**Discussion by gegoune** [hrsh7th/nvim-cmp#890] [Q&A] [answered]
Hey, I have noticed that if I omit a field from custom formatted window it gets way wider than it needs to be. For example, this is default with `formatting = {}` (I am using cmdline plugin here, but it's the same for main cmp's window):

<img width="165" alt="Screenshot 2022-04-12 at 21 39 54" src="https://user-images.githubusercontent.com/69750637/163031050-1240865d-36dc-459d-b8fd-56bb6fd6ac49.png">

So far so good. Let's add some kind icons (from wiki):
```lua
  formatting = {
    fields = { cmp.ItemField.Abbr, cmp.ItemField.Kind },
    format = function(_, vim_item)
      vim_item.kind = cmp_kinds[vim_item.kind] or ' '
      return vim_item
    end,
  },
```
and we get:

<img width="116" alt="Screenshot 2022-04-12 at 21 41 52" src="https://user-images.githubusercontent.com/69750637/163031368-88276af1-21d8-47c1-a5ee-6de44c071af7.png">

Still OK. Let's replace kind with source name coming from `entry` bundled with kind:

```lua
  formatting = {
    fields = { cmp.ItemField.Abbr, cmp.ItemField.Menu },
    format = function(entry, vim_item)
      vim_item.menu = ' '
        .. (({ nvim_lsp = 'lsp', cmp_git = 'git' })[entry.source.name] or entry.source.name)
        .. ': '
        .. vim_item.kind

      return vim_item
    end,
  },
```

<img width="315" alt="Screenshot 2022-04-12 at 21 43 34" src="https://user-images.githubusercontent.com/69750637/163031686-aac4ed77-08c5-4ff3-a72e-8d6d87df9f4a.png">

Not so good, window expands further than it really needs to. I was thinking that `menu` is breaking it somehow, but let add icons back in:

```lua
  formatting = {
    fields = { cmp.ItemField.Kind, cmp.ItemField.Abbr, cmp.ItemField.Menu },
    format = function(entry, vim_item)
      vim_item.menu = ' '
        .. (({ nvim_lsp = 'lsp', cmp_git = 'git' })[entry.source.name] or entry.source.name)
        .. ': '
        .. vim_item.kind

      vim_item.kind = cmp_kinds[vim_item.kind] or ' '

      return vim_item
    end,
  },

```

and everything is back to how it should be:

<img width="269" alt="Screenshot 2022-04-12 at 21 47 01" src="https://user-images.githubusercontent.com/69750637/163032255-4d905c20-df1e-4d0c-a270-68ed8ff1f3e4.png">

Why is abbr+kind displayed like that? @hrsh7th is that a bug or am I doing something utterly bananas here?
Thanks!

**Comment by GlekoMa** [comment]
This is a feature. Change items order will change the display order. See https://github.com/hrsh7th/nvim-cmp/issues/337
And this is the document (cmp.txt) description:
<img width="410" alt="fields" src="https://user-images.githubusercontent.com/97289978/211134204-ea5d0c77-7c4c-4dcb-be89-a87e268db300.png">

**Update：**
I'm sorry I misunderstood the description. 😥
@fitrh is right. `kind` and `abbr` display by default unless manually set.
Perhaps it would be more intuitive to leave it all to the `field` (not only specify the order but also behavior)?
I found an issue (https://github.com/hrsh7th/nvim-cmp/issues/1154#issue-1351431416) similar.

**Comment by fitrh** [comment, marked as answer]
```diff
  formatting = {
    fields = { cmp.ItemField.Abbr, cmp.ItemField.Menu },
    format = function(entry, vim_item)
      vim_item.menu = ' '
        .. (({ nvim_lsp = 'lsp', cmp_git = 'git' })[entry.source.name] or entry.source.name)
        .. ': '
        .. vim_item.kind

      return vim_item
    end,
  },
```

For this code to work as expected, the `vim_item.kind` need to set to `nil`

```diff
  formatting = {
    fields = { cmp.ItemField.Abbr, cmp.ItemField.Menu },
    format = function(entry, vim_item)
      vim_item.menu = ' '
        .. (({ nvim_lsp = 'lsp', cmp_git = 'git' })[entry.source.name] or entry.source.name)
        .. ': '
        .. vim_item.kind
+     vim_item.kind = nil
      return vim_item
    end,
  },
```

**Reply by gegoune** [reply]
I have no longer need to do this so won't test it, but it indeed looks like a solution. Thank you!
